Add shield iks01a3

avisconti opened the pull request:
x_nucleo_iks01a3 is a companion board that can be plugged through Arduino connectors on top
of Nucleo boards (such as nucleo_f411re).

This PR is adding shield support plus a couple of samples code, one for standard and one for sensorhub mode, plus proper documentation.
The PR is rebased on top of LPS22HH (#13062) and LSM6DSO (#13478) drivers.

avisconti committed:
boards: add support for shields x-nucleo-iks01a3

x-nucleo-iks01a3 shield is an arduino compatible companion board
which can be used on top of Nucleo standard boards. It extend
the Nucleo functionalities adding following MEMS sensor support:

    - LSM6DSO accel and gyro
    - LIS2MDL magnetometer
    - LIS2DW12 accel
    - LPS22HH pressure
    - HTS221 humidity

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
sample/shield: add support to x-nucleo-iks01a3 shield

Provide two basic example samples to test the x-nucleo-iks01a3 shield:

    - Standard (Mode 1)
    - SensorHub (Mode 2)

Signed-off-by: Armando Visconti <armando.visconti@st.com>

erwango commented:
@avisconti : I think we can remove tag "platform:STM32". This shield aims to be working on any board supporting arduino header, right?

avisconti commented:
> @avisconti : I think we can remove tag "platform:STM32". This shield aims to be working on any board supporting arduino header, right?

removed

ulfalizer commented:
Sorry about that false positive in a comment.

I really want to move ci-tools back into the main Zephyr repo. That way the whitelist could be modified in the same commit, without having to add yet another separate ad-hoc configuration file (which would have to support comments and stuff too).

There will be a pylint whitelist later too.

avisconti commented on boards/shields/x_nucleo_iks01a3/doc/index.rst:
```diff
@@ -0,0 +1,104 @@
+3-axis magnetometer, the LIS2DW12 3-axis accelerometer, the HTS221 humidity and
+temperature sensor, the LPS22HH pressure sensor, and the STTS751 temperature sensor.
+
+The X-NUCLEO-IKS01A3 interfaces with the STM32 microcontroller via the I²C pin,
```
ACK

avisconti commented on boards/shields/x_nucleo_iks01a3/doc/index.rst:
```diff
@@ -0,0 +1,104 @@
+Mode 1: Standard Mode
+=====================
+
+In standard I²C mode the two bus are connected together. As a consequence, all devices on the shield
```
thx

avisconti commented on boards/shields/x_nucleo_iks01a3/doc/index.rst:
```diff
@@ -0,0 +1,104 @@
+=====================
+
+In standard I²C mode the two bus are connected together. As a consequence, all devices on the shield
+resides on the same I²C bus and are accessible from the main board thru I²C bus.
```
thx

avisconti commented:
@dbkinder, thx for reviewing the board doc. I have fixed all your requested changes.
Have you taken a look to the two samples docs as well? Are they ok?

dbkinder reviewed (approved):
LGTM, thanks

avisconti commented:
I re-based the PR on top of the latest changes.
Please note that this PR depends from  #17112 and #17113.

avisconti commented:
@erwango 
I rebased it on top of master today.
The docuemntation has already been reviewed by dbkinder, but it would be nice if you could
take a look to the actual code itself. It should not take longer as it is quite similar to the other
sensor shields.

erwango commented on samples/shields/x_nucleo_iks01a3/sensorhub/src/main.c:
```diff
@@ -0,0 +1,295 @@
+#include <device.h>
+#include <sensor.h>
+#include <stdio.h>
+#include <misc/util.h>
```
Should be `sys/utils.h` according to on recent: https://github.com/zephyrproject-rtos/zephyr/pull/16540

avisconti replied:
ok, it makes sense.

avisconti replied:
Done. I did it for lps22hh and lsm6dso yesterday. But I forgot to apply the same here.
Thx!

erwango reviewed (changes requested):
Need to review headers inclusion due to recent change.

avisconti commented:
@nashif @galak 
Any further comment from your side?

MaureenHelm commented on boards/shields/x_nucleo_iks01a3/x_nucleo_iks01a3.overlay:
```diff
@@ -0,0 +1,41 @@
+	lps22hh@5d {
+		compatible = "st,lps22hh";
+		reg = <0x5d>;
+		drdy-gpios = <&gpiob 10 0>;
```
Can you look at the `arduino_header` introduced in #14057 and commit ce1e990c30f82be8a5c61a41dac1addf317c9538 , and make these gpios not specific to stm32 socs?

avisconti replied:
You mean the drdy gpios? Yes, excellent idea!
I will define arduino_header for nucleo F401re and F411re for testing purpose, and I will
change this shield overlay file to make use of it.

MaureenHelm commented on samples/shields/x_nucleo_iks01a3/sensorhub/README.rst:
```diff
@@ -0,0 +1,80 @@
+
+Please note that this sample can't be used with boards already supporting
+one of the sensors available on the shield (such as disco_l475_iot1) as zephyr
+does not yet support sensors multiple instances.
```
This is a little deceiving. Zephyr can certainly support multiple sensor instances, it's the specific drivers that don't actually implement multiple instances.

avisconti replied:
Actually this sentence was copied@paisted from other samples:
   https://docs.zephyrproject.org/latest/samples/shields/x_nucleo_iks01a1/README.html
https://docs.zephyrproject.org/latest/samples/shields/x_nucleo_iks01a2/README.html

But I can change it to "... as sensors multiple instances is not supported".

avisconti replied:
I changed it in this way:

 ```
Please note that this sample can't be used with boards already supporting
 one of the sensors available on the shield (such as disco_l475_iot1)
 as sensors multiple instances are not supported.

```
@MaureenHelm 
Pls check my english and tell me if it is better than before. Thank you.

avisconti committed:
boards/arm: Add arduino_header support to nucleo f401re and f411re

Add arduino_header support to nucleo f401re and f411re boards.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
@MaureenHelm  @erwango 
I added one commit for arduino_header support into nucleo f401re and f411re (it is the same mapping).
Then I changed the shield overlay file in order to have all drdy/irq gpios using the arduino_header.
Then rebased on top of latest master commits and retested both standard and sensorhub samples in both f411 and f401.
